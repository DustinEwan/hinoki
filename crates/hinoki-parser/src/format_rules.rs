use crate::Rule;
use pest::error::Error;

pub fn format_parse_error(err: Error<Rule>) -> Error<Rule> {
    err.renamed_rules(|rule| match *rule {
        Rule::EOI => "EOF".to_owned(),
        Rule::file => "import, struct, enum, fn, or EOF".to_owned(),

        // Include other rules above that aren't straightforward when hitting parsing errors!
        other => format!("{:?}", other),
    })
}
