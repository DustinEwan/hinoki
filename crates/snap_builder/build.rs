use std::{
    env,
    fs::{self, DirEntry},
    path::Path,
};

const SCRIPT_PATH: &str = "../hinoki-parser/testdata/snapshots";

fn main() {
    println!("cargo:rerun-if-changed={}", SCRIPT_PATH);
    let mut test_list_string = String::from("static TEST_FILES: [&str;");

    println!("{}", env::current_dir().unwrap().to_str().unwrap());
    let dir = fs::read_dir(SCRIPT_PATH);

    let mut filenames: Vec<String> = vec![];

    match dir {
        Ok(files) => {
            let files = files.flatten().collect::<Vec<DirEntry>>();
            let num_files = files.len();
            test_list_string.push_str(&format!("{}] = [\n", num_files).to_owned());

            files.into_iter().for_each(|file| {
                let filename = file.file_name();
                let filename = filename.to_str().unwrap();
                let filename = filename.replace(".hi", "");

                println!("Found file: {}", filename);

                filenames.push(format!("\t\"{}\"", filename).to_owned());
            });
        }
        Err(e) => panic!("{}", e),
    };

    test_list_string.push_str(&filenames.join(",\n"));
    test_list_string.push_str("\n];\n");

    println!("{}", test_list_string);

    let out_dir = env::var("OUT_DIR").unwrap();
    let dest_path = Path::new(&out_dir).join("testfiles.rs");
    let _ = fs::write(dest_path, test_list_string);
}
