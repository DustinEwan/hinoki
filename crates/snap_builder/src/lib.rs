extern crate proc_macro;
use proc_macro::TokenStream;

include!(concat!(env!("OUT_DIR"), "/testfiles.rs"));

#[proc_macro]
pub fn build_snaps(_item: TokenStream) -> TokenStream {
    TEST_FILES
        .iter()
        .map(|f| format!("snap!({});", f))
        .collect::<Vec<String>>()
        .join("\n")
        .parse()
        .unwrap()
}
