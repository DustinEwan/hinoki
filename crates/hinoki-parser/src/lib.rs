mod format_rules;

use once_cell::sync::OnceCell;
use tracing_subscriber::{fmt::FormattedFields, util::SubscriberInitExt};

use pest::Parser;
use pest::{error::Error, pratt_parser::PrattParser};
use pest_derive::Parser;

use crate::format_rules::format_parse_error;

#[derive(Parser)]
#[grammar = "./hinoki.pest"]
struct HinokiParser;

// lazy_static::lazy_static! {
//     static ref PRATT_PARSER: PrattParser<Rule> = {
//         use pest::pratt_parser::{Assoc::*, Op};
//         use Rule::*;
//
//         // Precedence is defined lowest to highest
//         PrattParser::new()
//             .op(Op::infix(xor, Left))
//             .op(Op::infix(bitwise_or, Left))
//             .op(Op::infix(bitwise_and, Left))
//             .op(Op::infix(not, Left))
//             .op(Op::infix(less_than, Left) | Op::infix(greater_than, Left) | Op::infix(equal, Left) | Op::infix(less_than_or_equal, Left) | Op::infix(greater_than_or_equal, Left) | Op::infix(not_equal, Left) )
//             .op(Op::infix(shift_left, Left) | Op::infix(shift_right, Left))
//             .op(Op::infix(add, Left) | Op::infix(subtract, Left))
//             .op(Op::infix(modulus, Left))
//             .op(Op::infix(multiply, Left) | Op::infix(divide, Left))
//             .op(Op::infix(power, Left))
//     };
// }

pub struct Program {
    pub instructions: Vec<Instruction>,
}

pub enum Instruction {
    Statement(Statement),
    Expression,
}

pub enum Statement {
    Declaration(DeclarationStmt),
    Assignment(AssignmentStmt),
    Return(ReturnStmt),
    Expression(Expression),
}

pub struct DeclarationStmt {}
pub struct AssignmentStmt {}
pub struct ReturnStmt {}

pub enum Expression {
    Integer(i64),
    Float(f64),
    Binary(BinaryExpr),
}

pub struct BinaryExpr {
    lhs: Box<Expression>,
    op: Op,
    rhs: Box<Expression>,
}

pub enum Op {
    Add,
    Subtract,
    Multiply,
    Divide,
    Mod,
    Power,

    ShiftLeft,
    ShiftRight,
    BitwiseAnd,
    BitwiseOr,
    Xor,
    Not,

    LessThan,
    GreaterThan,
    Equal,
    LessThanOrEqual,
    GreaterThanOrEqual,
    NotEqual,
}

// pub fn parse_expr(pairs: Pairs<Rule>) -> Expression {
//     PRATT_PARSER
//         .map_primary(|primary| match primary.as_rule() {
//             Rule::
//         })
// }

fn snapshot_parsing(input: &str) -> String {
    let file = HinokiParser::parse(Rule::file, input);

    match file {
        Ok(mut file) => format!("{:#?}", file.next().unwrap()),
        Err(e) => {
            let formatted_err = format_parse_error(e.clone());
            println!("{}", e);
            format!("{:#?}\n\n{}", e, formatted_err)
        }
    }
}

pub fn setup_trace() {
    static INSTANCE: OnceCell<()> = OnceCell::new();
    INSTANCE.get_or_init(|| {
        tracing_subscriber::fmt()
            .with_max_level(tracing::Level::TRACE)
            .without_time()
            .with_target(false)
            .finish()
            .init();
    });
}

#[cfg(test)]
mod test {
    use snap_builder::build_snaps;

    use crate::*;

    macro_rules! snap {
        ($name: tt) => {
            #[test]
            fn $name() {
                setup_trace();

                let contents =
                    include_str!(concat!("../testdata/snapshots/", stringify!($name), ".hi"));
                let mut settings = insta::Settings::clone_current();

                settings.set_snapshot_path("../testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_parsing(contents));
                });
            }
        };
    }

    build_snaps!();
    // snap!(test_declaration, "simple_arithmetic.hi");
}
