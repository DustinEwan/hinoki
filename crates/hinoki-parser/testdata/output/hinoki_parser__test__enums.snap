---
source: crates/hinoki-parser/src/lib.rs
expression: snapshot_parsing(contents)
---
Pair {
    rule: file,
    span: Span {
        str: "\nenum Basic {\n\tPass,\n\tFail\n}\n\nenum Option<T> {\n\tSome(T),\n\tNone,\n}\n\nenum Result<T, E> {\n\tOk(T),\n\tErr(E)\n}\n\nenum MultiDimVec {\n\tDim(float),\n\tVec2D(float, float),\n\tVec3D(float, float, float),\n}\n",
        start: 0,
        end: 191,
    },
    inner: [
        Pair {
            rule: enum_definition,
            span: Span {
                str: "enum Basic {\n\tPass,\n\tFail\n}",
                start: 1,
                end: 28,
            },
            inner: [
                Pair {
                    rule: enum,
                    span: Span {
                        str: "enum",
                        start: 1,
                        end: 5,
                    },
                    inner: [],
                },
                Pair {
                    rule: ident,
                    span: Span {
                        str: "Basic",
                        start: 6,
                        end: 11,
                    },
                    inner: [],
                },
                Pair {
                    rule: enum_variant,
                    span: Span {
                        str: "Pass",
                        start: 15,
                        end: 19,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "Pass",
                                start: 15,
                                end: 19,
                            },
                            inner: [],
                        },
                    ],
                },
                Pair {
                    rule: enum_variant,
                    span: Span {
                        str: "Fail\n",
                        start: 22,
                        end: 27,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "Fail",
                                start: 22,
                                end: 26,
                            },
                            inner: [],
                        },
                    ],
                },
            ],
        },
        Pair {
            rule: enum_definition,
            span: Span {
                str: "enum Option<T> {\n\tSome(T),\n\tNone,\n}",
                start: 30,
                end: 65,
            },
            inner: [
                Pair {
                    rule: enum,
                    span: Span {
                        str: "enum",
                        start: 30,
                        end: 34,
                    },
                    inner: [],
                },
                Pair {
                    rule: ident,
                    span: Span {
                        str: "Option",
                        start: 35,
                        end: 41,
                    },
                    inner: [],
                },
                Pair {
                    rule: type_generics_definition,
                    span: Span {
                        str: "<T>",
                        start: 41,
                        end: 44,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "T",
                                start: 42,
                                end: 43,
                            },
                            inner: [],
                        },
                    ],
                },
                Pair {
                    rule: enum_variant,
                    span: Span {
                        str: "Some(T)",
                        start: 48,
                        end: 55,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "Some",
                                start: 48,
                                end: 52,
                            },
                            inner: [],
                        },
                        Pair {
                            rule: type,
                            span: Span {
                                str: "T",
                                start: 53,
                                end: 54,
                            },
                            inner: [
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "T",
                                        start: 53,
                                        end: 54,
                                    },
                                    inner: [],
                                },
                            ],
                        },
                    ],
                },
                Pair {
                    rule: enum_variant,
                    span: Span {
                        str: "None",
                        start: 58,
                        end: 62,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "None",
                                start: 58,
                                end: 62,
                            },
                            inner: [],
                        },
                    ],
                },
            ],
        },
        Pair {
            rule: enum_definition,
            span: Span {
                str: "enum Result<T, E> {\n\tOk(T),\n\tErr(E)\n}",
                start: 67,
                end: 104,
            },
            inner: [
                Pair {
                    rule: enum,
                    span: Span {
                        str: "enum",
                        start: 67,
                        end: 71,
                    },
                    inner: [],
                },
                Pair {
                    rule: ident,
                    span: Span {
                        str: "Result",
                        start: 72,
                        end: 78,
                    },
                    inner: [],
                },
                Pair {
                    rule: type_generics_definition,
                    span: Span {
                        str: "<T, E>",
                        start: 78,
                        end: 84,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "T",
                                start: 79,
                                end: 80,
                            },
                            inner: [],
                        },
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "E",
                                start: 82,
                                end: 83,
                            },
                            inner: [],
                        },
                    ],
                },
                Pair {
                    rule: enum_variant,
                    span: Span {
                        str: "Ok(T)",
                        start: 88,
                        end: 93,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "Ok",
                                start: 88,
                                end: 90,
                            },
                            inner: [],
                        },
                        Pair {
                            rule: type,
                            span: Span {
                                str: "T",
                                start: 91,
                                end: 92,
                            },
                            inner: [
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "T",
                                        start: 91,
                                        end: 92,
                                    },
                                    inner: [],
                                },
                            ],
                        },
                    ],
                },
                Pair {
                    rule: enum_variant,
                    span: Span {
                        str: "Err(E)",
                        start: 96,
                        end: 102,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "Err",
                                start: 96,
                                end: 99,
                            },
                            inner: [],
                        },
                        Pair {
                            rule: type,
                            span: Span {
                                str: "E",
                                start: 100,
                                end: 101,
                            },
                            inner: [
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "E",
                                        start: 100,
                                        end: 101,
                                    },
                                    inner: [],
                                },
                            ],
                        },
                    ],
                },
            ],
        },
        Pair {
            rule: enum_definition,
            span: Span {
                str: "enum MultiDimVec {\n\tDim(float),\n\tVec2D(float, float),\n\tVec3D(float, float, float),\n}",
                start: 106,
                end: 190,
            },
            inner: [
                Pair {
                    rule: enum,
                    span: Span {
                        str: "enum",
                        start: 106,
                        end: 110,
                    },
                    inner: [],
                },
                Pair {
                    rule: ident,
                    span: Span {
                        str: "MultiDimVec",
                        start: 111,
                        end: 122,
                    },
                    inner: [],
                },
                Pair {
                    rule: enum_variant,
                    span: Span {
                        str: "Dim(float)",
                        start: 126,
                        end: 136,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "Dim",
                                start: 126,
                                end: 129,
                            },
                            inner: [],
                        },
                        Pair {
                            rule: type,
                            span: Span {
                                str: "float",
                                start: 130,
                                end: 135,
                            },
                            inner: [
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "float",
                                        start: 130,
                                        end: 135,
                                    },
                                    inner: [],
                                },
                            ],
                        },
                    ],
                },
                Pair {
                    rule: enum_variant,
                    span: Span {
                        str: "Vec2D(float, float)",
                        start: 139,
                        end: 158,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "Vec2D",
                                start: 139,
                                end: 144,
                            },
                            inner: [],
                        },
                        Pair {
                            rule: type,
                            span: Span {
                                str: "float",
                                start: 145,
                                end: 150,
                            },
                            inner: [
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "float",
                                        start: 145,
                                        end: 150,
                                    },
                                    inner: [],
                                },
                            ],
                        },
                        Pair {
                            rule: type,
                            span: Span {
                                str: "float",
                                start: 152,
                                end: 157,
                            },
                            inner: [
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "float",
                                        start: 152,
                                        end: 157,
                                    },
                                    inner: [],
                                },
                            ],
                        },
                    ],
                },
                Pair {
                    rule: enum_variant,
                    span: Span {
                        str: "Vec3D(float, float, float)",
                        start: 161,
                        end: 187,
                    },
                    inner: [
                        Pair {
                            rule: ident,
                            span: Span {
                                str: "Vec3D",
                                start: 161,
                                end: 166,
                            },
                            inner: [],
                        },
                        Pair {
                            rule: type,
                            span: Span {
                                str: "float",
                                start: 167,
                                end: 172,
                            },
                            inner: [
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "float",
                                        start: 167,
                                        end: 172,
                                    },
                                    inner: [],
                                },
                            ],
                        },
                        Pair {
                            rule: type,
                            span: Span {
                                str: "float",
                                start: 174,
                                end: 179,
                            },
                            inner: [
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "float",
                                        start: 174,
                                        end: 179,
                                    },
                                    inner: [],
                                },
                            ],
                        },
                        Pair {
                            rule: type,
                            span: Span {
                                str: "float",
                                start: 181,
                                end: 186,
                            },
                            inner: [
                                Pair {
                                    rule: ident,
                                    span: Span {
                                        str: "float",
                                        start: 181,
                                        end: 186,
                                    },
                                    inner: [],
                                },
                            ],
                        },
                    ],
                },
            ],
        },
        Pair {
            rule: EOI,
            span: Span {
                str: "",
                start: 191,
                end: 191,
            },
            inner: [],
        },
    ],
}
